configuration.md
---
# !!! DO NOT MODIFY !!!
# This file is auto-generated by the gendoc tool based on the source code of the app.
description: This section describes the configuration parameters and their types for INX-Coordinator.
keywords:
- IOTA Node 
- Hornet Node
- Coordinator
- Configuration
- JSON
- Customize
- Config
- reference
---


# Core Configuration

INX-Coordinator uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `inx-coordinator` executable.

For example:
```bash
inx-coordinator -c config_defaults.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
inx-coordinator -h --full
```

## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "5m"          |
| [log](#app_shutdown_log) | Configuration for log                                                                                  | object |               |

### <a id="app_shutdown_log"></a> Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "5m",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name                                     | Description                                                                 | Type    | Default value |
| ---------------------------------------- | --------------------------------------------------------------------------- | ------- | ------------- |
| level                                    | The minimum enabled logging level                                           | string  | "info"        |
| disableCaller                            | Stops annotating logs with the calling function's file name and line number | boolean | true          |
| disableStacktrace                        | Disables automatic stacktrace capturing                                     | boolean | false         |
| stacktraceLevel                          | The level stacktraces are captured and above                                | string  | "panic"       |
| encoding                                 | The logger's encoding (options: "json", "console")                          | string  | "console"     |
| [encodingConfig](#logger_encodingconfig) | Configuration for encodingConfig                                            | object  |               |
| outputPaths                              | A list of URLs, file paths or stdout/stderr to write logging output to      | array   | stdout        |
| disableEvents                            | Prevents log messages from being raced as events                            | boolean | true          |

### <a id="logger_encodingconfig"></a> EncodingConfig

| Name        | Description                                                                                                | Type   | Default value |
| ----------- | ---------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| timeEncoder | Sets the logger's timestamp encoding. (options: "nanos", "millis", "iso8601", "rfc3339" and "rfc3339nano") | string | "rfc3339"     |

Example:

```json
  {
    "logger": {
      "level": "info",
      "disableCaller": true,
      "disableStacktrace": false,
      "stacktraceLevel": "panic",
      "encoding": "console",
      "encodingConfig": {
        "timeEncoder": "rfc3339"
      },
      "outputPaths": [
        "stdout"
      ],
      "disableEvents": true
    }
  }
```

## <a id="inx"></a> 3. INX

| Name                  | Description                                                                                        | Type   | Default value    |
| --------------------- | -------------------------------------------------------------------------------------------------- | ------ | ---------------- |
| address               | The INX address to which to connect to                                                             | string | "localhost:9029" |
| maxConnectionAttempts | The amount of times the connection to INX will be attempted before it fails (1 attempt per second) | uint   | 30               |
| targetNetworkName     | The network name on which the node should operate on (optional)                                    | string | ""               |

Example:

```json
  {
    "inx": {
      "address": "localhost:9029",
      "maxConnectionAttempts": 30,
      "targetNetworkName": ""
    }
  }
```

## <a id="coordinator"></a> 4. Coordinator

| Name                                      | Description                                                                                                  | Type    | Default value       |
| ----------------------------------------- | ------------------------------------------------------------------------------------------------------------ | ------- | ------------------- |
| stateFilePath                             | The path to the state file of the coordinator                                                                | string  | "coordinator.state" |
| interval                                  | The interval milestones are issued                                                                           | string  | "5s"                |
| milestoneTimeout                          | The duration after which an event is triggered if no new milestones are received                             | string  | "30s"               |
| [signing](#coordinator_signing)           | Configuration for signing                                                                                    | object  |                     |
| [quorum](#coordinator_quorum)             | Configuration for quorum                                                                                     | object  |                     |
| [checkpoints](#coordinator_checkpoints)   | Configuration for checkpoints                                                                                | object  |                     |
| [tipsel](#coordinator_tipsel)             | Configuration for Tipselection                                                                               | object  |                     |
| [blockBackups](#coordinator_blockbackups) | Configuration for blockBackups                                                                               | object  |                     |
| debugFakeMilestoneTimestamps              | Whether the coordinator will fake timestamps of milestones if the interval is below 1s (use for tests only!) | boolean | false               |

### <a id="coordinator_signing"></a> Signing

| Name          | Description                                                                    | Type   | Default value     |
| ------------- | ------------------------------------------------------------------------------ | ------ | ----------------- |
| provider      | The signing provider the coordinator uses to sign a milestone (local/remote)   | string | "local"           |
| remoteAddress | The address of the remote signing provider (insecure connection!)              | string | "localhost:12345" |
| retryTimeout  | Defines the timeout between signing retries                                    | string | "2s"              |
| retryAmount   | Defines the number of signing retries to perform before shutting down the node | int    | 10                |

### <a id="coordinator_quorum"></a> Quorum

| Name    | Description                                                                                    | Type    | Default value     |
| ------- | ---------------------------------------------------------------------------------------------- | ------- | ----------------- |
| enabled | Whether the coordinator quorum is enabled                                                      | boolean | false             |
| timeout | The timeout until a node in the quorum must have answered                                      | string  | "2s"              |
| groups  | Defines the quorum groups used to ask other nodes for correct ledger state of the coordinator. | object  | see example below |

### <a id="coordinator_checkpoints"></a> Checkpoints

| Name             | Description                                                                                                       | Type | Default value |
| ---------------- | ----------------------------------------------------------------------------------------------------------------- | ---- | ------------- |
| maxTrackedBlocks | Maximum amount of known blocks for milestone tipselection. If this limit is exceeded, a new checkpoint is issued. | int  | 10000         |

### <a id="coordinator_tipsel"></a> Tipselection

| Name                                         | Description                                                                                                                                                                    | Type   | Default value |
| -------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| minHeaviestBranchUnreferencedBlocksThreshold | Minimum threshold of unreferenced blocks in the heaviest branch                                                                                                                | int    | 20            |
| maxHeaviestBranchTipsPerCheckpoint           | Maximum amount of checkpoint blocks with heaviest branch tips that are picked if the heaviest branch is not below 'MinHeaviestBranchUnreferencedBlocksThreshold' before        | int    | 10            |
| randomTipsPerCheckpoint                      | Amount of checkpoint blocks with random tips that are picked if a checkpoint is issued and at least one heaviest branch tip was found, otherwise no random tips will be picked | int    | 3             |
| heaviestBranchSelectionTimeout               | The maximum duration to select the heaviest branch tips                                                                                                                        | string | "100ms"       |

### <a id="coordinator_blockbackups"></a> BlockBackups

| Name       | Description                                                                                                          | Type    | Default value   |
| ---------- | -------------------------------------------------------------------------------------------------------------------- | ------- | --------------- |
| enabled    | Whether all blocks that are issued by the coordinator should be stored to disk before being submitted to the network | boolean | true            |
| folderPath | The path to the folder where block backups are stored                                                                | string  | "block_backups" |

Example:

```json
  {
    "coordinator": {
      "stateFilePath": "coordinator.state",
      "interval": "5s",
      "milestoneTimeout": "30s",
      "signing": {
        "provider": "local",
        "remoteAddress": "localhost:12345",
        "retryTimeout": "2s",
        "retryAmount": 10
      },
      "quorum": {
        "enabled": false,
        "timeout": "2s",
        "groups": {}
      },
      "checkpoints": {
        "maxTrackedBlocks": 10000
      },
      "tipsel": {
        "minHeaviestBranchUnreferencedBlocksThreshold": 20,
        "maxHeaviestBranchTipsPerCheckpoint": 10,
        "randomTipsPerCheckpoint": 3,
        "heaviestBranchSelectionTimeout": "100ms"
      },
      "blockBackups": {
        "enabled": true,
        "folderPath": "block_backups"
      },
      "debugFakeMilestoneTimestamps": false
    }
  }
```

## <a id="migrator"></a> 5. Migrator

| Name                | Description                                                                                                           | Type    | Default value    |
| ------------------- | --------------------------------------------------------------------------------------------------------------------- | ------- | ---------------- |
| enabled             | Whether the migrator plugin is enabled                                                                                | boolean | false            |
| stateFilePath       | Path to the state file of the migrator                                                                                | string  | "migrator.state" |
| receiptMaxEntries   | The max amount of entries to embed within a receipt                                                                   | int     | 110              |
| queryCooldownPeriod | The cooldown period for the service to ask for new data from the legacy node in case the migrator encounters an error | string  | "5s"             |

Example:

```json
  {
    "migrator": {
      "enabled": false,
      "stateFilePath": "migrator.state",
      "receiptMaxEntries": 110,
      "queryCooldownPeriod": "5s"
    }
  }
```

## <a id="receipts"></a> 6. Receipts

| Name                             | Description                 | Type   | Default value |
| -------------------------------- | --------------------------- | ------ | ------------- |
| [validator](#receipts_validator) | Configuration for validator | object |               |

### <a id="receipts_validator"></a> Validator

| Name                                           | Description                   | Type   | Default value |
| ---------------------------------------------- | ----------------------------- | ------ | ------------- |
| [api](#receipts_validator_api)                 | Configuration for API         | object |               |
| [coordinator](#receipts_validator_coordinator) | Configuration for coordinator | object |               |

### <a id="receipts_validator_api"></a> API

| Name    | Description                                                              | Type   | Default value            |
| ------- | ------------------------------------------------------------------------ | ------ | ------------------------ |
| address | Address of the legacy node API to query for white-flag confirmation data | string | "http://localhost:14266" |
| timeout | Timeout of API calls                                                     | string | "5s"                     |

### <a id="receipts_validator_coordinator"></a> Coordinator

| Name            | Description                                 | Type   | Default value                                                                       |
| --------------- | ------------------------------------------- | ------ | ----------------------------------------------------------------------------------- |
| address         | Address of the legacy coordinator           | string | "UDYXTZBE9GZGPM9SSQV9LTZNDLJIZMPUVVXYXFYVBLIEUHLSEWFTKZZLXYRHHWVQV9MNNX9KZC9D9UZWZ" |
| merkleTreeDepth | Depth of the Merkle tree of the coordinator | int    | 24                                                                                  |

Example:

```json
  {
    "receipts": {
      "validator": {
        "api": {
          "address": "http://localhost:14266",
          "timeout": "5s"
        },
        "coordinator": {
          "address": "UDYXTZBE9GZGPM9SSQV9LTZNDLJIZMPUVVXYXFYVBLIEUHLSEWFTKZZLXYRHHWVQV9MNNX9KZC9D9UZWZ",
          "merkleTreeDepth": 24
        }
      }
    }
  }
```

## <a id="profiling"></a> 7. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling component is enabled        | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```



components/migrator/params.go
package migrator

import (
	"time"

	"github.com/iotaledger/hive.go/app"
	"github.com/iotaledger/inx-coordinator/pkg/migrator"
)

// ParametersMigrator contains the definition of the parameters used by Migrator.
type ParametersMigrator struct {
	// Enabled defines whether the migrator plugin is enabled.
	Enabled bool `default:"false" usage:"whether the migrator plugin is enabled"`
	// StateFilePath defines the path to the state file of the migrator.
	StateFilePath string `default:"migrator.state" usage:"path to the state file of the migrator"`
	// ReceiptMaxEntries defines the max amount of entries to embed within a receipt.
	ReceiptMaxEntries int `usage:"the max amount of entries to embed within a receipt"`
	// QueryCooldownPeriod defines the cooldown period for the service to ask for new data from the legacy node in case the migrator encounters an error.
	QueryCooldownPeriod time.Duration `default:"5s" usage:"the cooldown period for the service to ask for new data from the legacy node in case the migrator encounters an error"`
}

var ParamsMigrator = &ParametersMigrator{
	ReceiptMaxEntries: migrator.SensibleMaxEntriesCount,
}

// ParametersReceipts contains the definition of the parameters used by Receipts.
type ParametersReceipts struct {
	Validator struct {
		API struct {
			// Address defines the address of the legacy node API to query for white-flag confirmation data.
			Address string `default:"http://localhost:14266" usage:"address of the legacy node API to query for white-flag confirmation data"`
			// Address defines the timeout of API calls.
			Timeout time.Duration `default:"5s" usage:"timeout of API calls"`
		} `name:"api"`
		Coordinator struct {
			// Address defines the address of the legacy coordinator.
			Address string `default:"UDYXTZBE9GZGPM9SSQV9LTZNDLJIZMPUVVXYXFYVBLIEUHLSEWFTKZZLXYRHHWVQV9MNNX9KZC9D9UZWZ" usage:"address of the legacy coordinator"`
			// MerkleTreeDepth defines the depth of the Merkle tree of the coordinator.
			MerkleTreeDepth int `default:"24" usage:"depth of the Merkle tree of the coordinator"`
		}
	}
}

var ParamsReceipts = &ParametersReceipts{}

var params = &app.ComponentParams{
	Params: map[string]any{
		"migrator": ParamsMigrator,
		"receipts": ParamsReceipts,
	},
	Masked: nil,
}
